src/Microsoft.AspNetCore.Blazor.Server/AutoRebuild/AutoRebuildExtensions.cs

```diff
@@ -0,0 +1,109 @@
+﻿// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using Microsoft.AspNetCore.Blazor.Server;
+using Microsoft.AspNetCore.Blazor.Server.AutoRebuild;
+using System;
+using System.IO;
+using System.Linq;
+using System.Threading.Tasks;
+
+namespace Microsoft.AspNetCore.Builder
+{
+    internal static class AutoRebuildExtensions
+    {
+        // Note that we don't need to watch typical static-file extensions (.css, .js, etc.)
+        // because anything in wwwroot is just served directly from disk on each reload. But
+        // as a special case, we do watch index.html because it needs compilation.
+        // TODO: Make the set of extensions and exclusions configurable in csproj
+        private static string[] _includedSuffixes = new[] { ".cs", ".cshtml", "index.html" };
+        private static string[] _excludedDirectories = new[] { "obj", "bin" };
+
+        public static void UseAutoRebuild(this IApplicationBuilder appBuilder, BlazorConfig config)
+        {
+            // Currently this only supports VS for Windows. Later on we can add
+            // an IRebuildService implementation for VS for Mac, etc.
+            if (!VSForWindowsRebuildService.TryCreate(out var rebuildService))
+            {
+                return; // You're not on Windows, or you didn't launch this process from VS
+            }
+
+            var currentCompilationTask = Task.CompletedTask;
+            var compilationTaskMustBeRefreshedIfNotBuiltSince = (DateTime?)null;
+
+            WatchFileSystem(config, () =>
+            {
+                // Don't start the recompilation immediately. We only start it when the next
+                // HTTP request arrives, because it's annoying if the IDE is constantly rebuilding
+                // when you're making changes to multiple files and aren't ready to reload
+                // in the browser yet.
+                compilationTaskMustBeRefreshedIfNotBuiltSince = DateTime.Now;
+            });
+
+            appBuilder.Use(async (context, next) =>
+            {
+                try
+                {
+                    if (compilationTaskMustBeRefreshedIfNotBuiltSince.HasValue)
+                    {
+                        var rebuildIfNotBuiltSince = compilationTaskMustBeRefreshedIfNotBuiltSince.Value;
```

Comment:
There are still numerous race conditions here, I don't think it's a big problem because the exceptions are going to be handled, but any truly correct solution is going to involve a lock or loop since you have two resources.